src/pages/spawn/spawnVolume/spawnVolumeButton/SpawnVolumeModal.tsx

```diff
@@ -34,111 +26,101 @@ export const SpawnVolumeModal: React.VFC<SpawnVolumeModalProps> = ({
       );
     },
     refetchQueries: ["MyVolumes"],
   });
 
-  useEffect(() => {
-    if (visible) {
-      dispatch({ type: "reset" });
-    }
-  }, [visible, dispatch]);
-
   const spawnVolume = () => {
-    const mutationVars = { ...state };
-    if (mutationVars.noExpiration === true) {
-      delete mutationVars.expiration;
-    }
-    if (mutationVars.host === "") {
-      delete mutationVars.host;
-    }
-    const variables = { SpawnVolumeInput: mutationVars };
-    spawnAnalytics.sendEvent({ name: "Spawned a volume", params: variables });
-    spawnVolumeMutation({ variables });
+    const mutationInput = formToGql({ formData: formState });
+    spawnAnalytics.sendEvent({
+      name: "Spawned a volume",
+      params: mutationInput,
+    });
+    spawnVolumeMutation({
+      variables: { SpawnVolumeInput: mutationInput },
+    });
   };
-  const volumeLimit = spruceConfig?.providers?.aws?.maxVolumeSizePerUser;
-  const totalVolumeSize = volumesData?.myVolumes?.reduce(
-    (cnt, v) => cnt + v.size,
-    0
-  );
-  const maxSpawnableLimit =
-    volumeLimit - totalVolumeSize >= 0 ? volumeLimit - totalVolumeSize : 0;
 
-  useEffect(() => {
-    // Update the size input when we set a new max volume size limit
-    // If the max size limit is > 500 default to 500
-    dispatch({
-      type: "setSize",
-      data: maxSpawnableLimit > 500 ? 500 : maxSpawnableLimit,
-    });
-  }, [maxSpawnableLimit, totalVolumeSize, volumeLimit]);
+  const {
+    availabilityZones,
+    types,
+    hosts,
+    disableExpirationCheckbox,
+    noExpirationCheckboxTooltip,
+    maxSpawnableLimit,
+    loadingFormData,
+  } = useLoadFormData();
+
+  const [formState, setFormState] = useState<FormState>({});
+
+  const availableHosts = hosts
+    .filter(
+      ({ availabilityZone, status }) =>
+        availabilityZone ===
+          formState?.requiredVolumeInformation?.availabilityZone &&
+        (status === HostStatus.Running || status === HostStatus.Stopped)
+    )
+    .map(({ id, displayName }) => ({ id, displayName }))
+    .sort((a, b) => a.displayName.localeCompare(b.displayName));
+
+  const { schema, uiSchema } = getFormSchema({
+    maxSpawnableLimit,
+    availabilityZones,
+    types,
+    hosts: availableHosts,
+    disableExpirationCheckbox,
+    noExpirationCheckboxTooltip,
+  });
+
+  if (loadingFormData) {
+    return null;
+  }
 
   return (
     <ConfirmationModal
       title="Spawn New Volume"
       open={visible}
-      onCancel={onCancel}
-      buttonText={loadingSpawnVolume ? "Spawning Volume" : "Spawn"}
+      onCancel={() => {
+        closeModal();
+      }}
+      buttonText={loadingSpawnVolume ? "Spawning volume" : "Spawn"}
       onConfirm={spawnVolume}
       submitDisabled={
-        loadingSpawnVolume || !state.size || state.size > maxSpawnableLimit
+        loadingSpawnVolume ||
+        !formState?.requiredVolumeInformation?.volumeSize ||
+        formState?.requiredVolumeInformation?.volumeSize > maxSpawnableLimit
```

Comment:
RJSF's onChange function also includes an errors param that I like to use to enable/disable submission. It's handy because the form can be expanded and validated without needing to change the conditionals in use here. Here's a good example of it in use:

https://github.com/evergreen-ci/spruce/blob/5b97b24758e6213d6a621f09f76510a800b5c163/src/pages/task/taskTabs/buildBaronAndAnnotations/AddIssueModal/index.tsx#L108-L111